src/drivers/intel/cavs/mn.c

```diff
@@ -61,28 +85,49 @@ int mn_set_mclk(uint16_t mclk_id, uint32_t mclk_rate)
 	}
 
 	spin_lock(mn_lock);
-	mdivc = mn_reg_read(0x0);
 
-	/* Enable MCLK Divider */
-	mdivc |= 0x1;
+	mclk_sources[mclk_id] = MN_MCLK_SOURCE_NONE;
 
-	/* searching the smallest possible mclk source */
-	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
-		if (mclk_rate > ssp_freq[i].freq)
-			break;
+	if (is_mclk_source_in_use()) {
+		clk_index = mclk_source_clock;
 
-		if (ssp_freq[i].freq % mclk_rate == 0)
-			clk_index = i;
-	}
+		if (ssp_freq[clk_index].freq % mclk_rate != 0) {
```

Comment:
@jajanusz for my education, in the case that this is 0, are we going to overwrite the mdivr based on the mdvir_val computed for the mclk_rate for the current SSP port being configured. Is this intentional?

Reply:
MCLKs are not bound to SSP ports, they are independent. You have f.e. 6 SSP ports and 2 MCLKs. You configure which MCLK to use from topology with mclk_id in SSP configs. If you use some source for mclk_id==0 and then want to configure mclk_id==1, then you cannot switch source that is shared by both 0&1 and only option is to get requested value by configuring divider.

Going back to your question - if you configure mclk_id==0 with some rate and then do it again for mclk_id==0, then you may overwrite it, but I assumed driver knows what it's doing, do you think we should check it? I wanted to keep it backwards compatible, because doing mentioned check would require to first `free` mclk for current divider and could make some flows fail. You would have to be sure that you destroyed DAI before using it for next SSP+MCLK config (IDK how you do it in driver).

Anyway even if you want me to do that check I'd do it in another PR, this one has too many changes at this point and I want to add more.

Reply:
@ranj063 Did I answer your question?